src/Data/Bin/Bin0.agda

```diff
@@ -0,0 +1,655 @@
+suc≢0 {suc2* _} ()
+
+infixl 6 _+_
+
+_+_ : Op₂ Bin
+0#        + y         =  y
+x         + 0#        =  x
+(2suc x)  + (2suc y)  =  2suc (suc (x + y))
+                                       -- 2(1+x) + 2(1+y) =  2(1 + 1+x+y)
+(2suc x)  + (suc2* y) =  suc (2suc (x + y))
+                                       -- 2(1+x) + 1 + 2y =  1 + 2(1+x+y)
+(suc2* x) + (2suc y)  =  suc (2suc (x + y))
+(suc2* x) + (suc2* y) =  suc (suc2* (x + y))
+                              -- 1+2x + 1+2y =  2 + 2(x+y) =  1 + 1 + 2(x+y)
+
+x≢0⇒x+y≢0 :  {x : Bin} (y : Bin) → x ≢ 0# → x + y ≢ 0#
+x≢0⇒x+y≢0 {2suc _} 0# _    =  λ()
+x≢0⇒x+y≢0 {0#}     _  0≢0 =  contradiction refl 0≢0
+
+2* : Bin → Bin
+2* 0#        = 0#
+2* (2suc x)  = 2suc (suc2* x)
+                            -- 2(1+x) + 2(1+x) = 2(1+x + 1+x) =  2(1 + 1+2x)
+2* (suc2* x) = 2suc (2* x)
+
+pred : Bin → Bin
+pred 0#        =  0#
+pred (2suc x)  =  suc2* x   -- 2(1+x) - 1 =  1+2x
+pred (suc2* x) =  2* x      -- 1 + 2x -1  =  2x
+
+2suc-as∘ :  2suc ≗ 2* ∘ suc
+2suc-as∘ 0#        =  refl
+2suc-as∘ (2suc x)  =  begin
+  2suc (2suc x)        ≡⟨ cong 2suc (2suc-as∘ x) ⟩
+  2suc (2* (suc x))    ≡⟨⟩
+  2* (suc2* (suc x))   ≡⟨⟩
+  2* (suc (2suc x))    ∎
+
+2suc-as∘ (suc2* x) =  refl
+
+
+-- suc2*∘suc≗suc∘2suc :  suc2* ∘ suc ≗ suc ∘ 2suc   is by  const refl
+
+suc2*-as∘ :  suc2* ≗ suc ∘ 2*
+suc2*-as∘ 0#        = refl
+suc2*-as∘ (2suc x)  = refl
+suc2*-as∘ (suc2* x) = begin
+  suc2* (suc2* x)     ≡⟨ cong suc2* (suc2*-as∘ x) ⟩
+  suc2* (suc 2x)      ≡⟨⟩
+  suc (2suc 2x)       ≡⟨ cong suc (2suc-as∘ 2x) ⟩
+  suc (2* (suc 2x))   ≡⟨ cong (suc ∘ 2*) (sym (suc2*-as∘ x)) ⟩
+  suc (2* (suc2* x))  ∎
+  where
+  2x = 2* x
+
+pred∘suc :  pred ∘ suc ≗ id
+pred∘suc 0#        =  refl
+pred∘suc (2suc x)  =  sym (2suc-as∘ x)
+pred∘suc (suc2* x) =  refl
+
+suc∘pred :  (x : Bin) → x ≢ 0# → suc (pred x) ≡ x
+
+suc∘pred 0#        0≢0 =  contradiction refl 0≢0
+suc∘pred (2suc _)  _   =  refl
+suc∘pred (suc2* x) _   =  begin
+  suc (pred (suc2* x))   ≡⟨⟩
+  suc (2* x)             ≡⟨ sym (suc2*-as∘ x) ⟩
+  suc2* x                ∎
+
+
++-identityˡ : LeftIdentity 0# _+_
++-identityˡ _ = refl
+
++-identityʳ : RightIdentity 0# _+_
++-identityʳ 0#        = refl
++-identityʳ (2suc _)  = refl
++-identityʳ (suc2* _) = refl
+
+2^ :  ℕ → Bin
+2^ 0      =  1#
+2^ (1+ n) =  2* (2^ n)
+
+
+infixl 7 _*_
+
+_*_ : Op₂ Bin
+
+0#        * _         =  0#
+_         * 0#        =  0#
+(2suc x)  * (2suc y)  =  2* (2suc (x + (y + x * y)))
+(2suc x)  * (suc2* y) =  2suc (x + y * (2suc x))
+                      --
+                      -- 2(1+x) * (1+2y) =  2(1 + 2y + x + 2xy)
+                      --                 =  2(1 + x + y*2(1 + x))
+
+(suc2* x) * (2suc y)  =  2suc (y + x * (2suc y))
+(suc2* x) * (suc2* y) =  suc2* (x + y * (suc2* x))
+            --
+            -- (1 + 2x)(1 + 2y) =  1 + (2y + 2x + 4xy)
+            --                     1 + 2(x + y * (1 + 2x))
+
+
+infixl 8 _^_
+
+_^_ : Bin → ℕ → Bin
+_ ^ 0      =  1#
+x ^ (1+ n) =  x * (x ^ n)
+
```

Comment:
We should exploit the fact that we have `Bin` at hand and can thus implement an
efficient [exponentiation by squaring](https://en.wikipedia.org/wiki/Exponentiation_by_squaring) of type `_^_ : Bin → Bin → Bin`.

Reply:
Please, let us delay this subject to the end of the New Bin commit.

``_^_`` and its proofs need to be implemented via squaring uniformly for all Monoid instances 
(in particular,  for Nat, Integer, Bin, BinaryInteger).
For Nat, the generic ``_^_`` to be imported from Algebra.Properties.Monoid with substituting 
the parameter ``NatProp.*-monoid``. Similarly are imported the proofs for ``Nat._^_``.
For Integer, one will need to substitute the parameter value ``IntegerProp.*-monoid``,
and so on. 
The algorithm and proofs will be implemented once.

Reply:
Then we can simply drop the function if it's meant to be replaced later on?

Reply:
> Then we can simply drop the function if it's meant to be replaced later on?

Yes.